model_compression_toolkit/gptq/pytorch/gptq_training.py

```diff
@@ -106,20 +111,84 @@ def _get_total_grad_steps():
-        self.hessian_score_per_image_per_layer = None    # for sample-layer attention
         if self.use_sample_layer_attention:
             # normalization is currently not supported, make sure the config reflects it.
             if hessian_cfg.norm_scores or hessian_cfg.log_norm or hessian_cfg.scale_log_norm:
                 raise NotImplementedError()
-            # Per sample hessian scores are calculated on-demand during the training loop
-            self.hessian_score_per_image_per_layer = {}
+            self.train_dataloader = self._prepare_train_dataloader_sla(representative_data_gen)
         else:
-            self.hessian_score_per_layer = to_torch_tensor(self.compute_hessian_based_weights())
+            self.train_dataloader = self._prepare_train_dataloader_for_non_sla(representative_data_gen)
 
         self.reg_func = get_regularization(self.gptq_config, _get_total_grad_steps)
 
+    def _prepare_train_dataloader_sla(self, data_gen_fn: Callable[[], Generator]) -> DataLoader:
+        """
+        Computes Sample-Layer Attention score and builds a train dataloader.
+
+        Args:
+            data_gen_fn: factory for representative dataset generator.
+
+        Returns:
+            PyTorch dataloader yielding three outputs - samples, weights for the distillation loss and
+              weights for regularization.
+        """
+        fixed_dataset = FixedDatasetFromGenerator(data_gen_fn)
+        orig_batch_size = fixed_dataset.orig_batch_size
+        # compute hessians for the whole dataset
+        hess_data_loader = DataLoader(fixed_dataset,
+                                      batch_size=self.gptq_config.hessian_weights_config.hessian_batch_size,
+                                      shuffle=False)
+        request = self._build_hessian_request(granularity=HessianScoresGranularity.PER_OUTPUT_CHANNEL,
+                                              data_loader=hess_data_loader,
+                                              n_samples=None)
+        layers_hessians = self.hessian_service.fetch_hessian(request, force_compute=True)
+
+        # compute sla score defined as max over channels
+        layers_hessians = {layer: to_torch_tensor(hess.max(1)) for layer, hess in layers_hessians.items()}
+
+        # build train dataset and dataloader
+        hessians_tensor = torch.stack([layers_hessians[layer.name] for layer in self.compare_points], dim=1)    # samples X layers
+        assert hessians_tensor.shape[1] == len(self.compare_points)
+        loss_weights = list(hessians_tensor)
+        sla_train_dataset = FixedSampleInfoDataset(fixed_dataset.samples, loss_weights)
+
+        reg_weights = hessians_tensor.mean(dim=0)
+        # use collate to add a single value to each batch
+        collate_fn = get_collate_fn_with_extra_outputs(reg_weights)
+
+        return DataLoader(sla_train_dataset, batch_size=orig_batch_size, shuffle=True, collate_fn=collate_fn)
```

Comment:
I see now the "shuffle" flag so it probably answers my previous question regarding the shuffle, but I'm keeping it open just to make sure